### PR TITLE
fix: Unify python dependency installation and update to vulnerability free versions

### PR DESCRIPTION
**Pull request by jparsai**
# Description
- [x] Resolved vulnerabilities and used updated versions of the dependencies.
- [x] Moved any test specific packages from requirements.in to tests/requirements.in
- [x] Moved any direct pip installations from Dockerfile.* to requirements.in
- [x] Provided relative versioning in requirements.in & tests/requirements.in
- [x] Generated requirements.txt and tests/requirements.txt using pip compiler
- [x] Updated setup.py to use requirement.in

## Issue
APPAI-1406/1409

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -21,12 +20,8 @@ COPY requirements.txt /tmp/f8a_worker/
 RUN cd /tmp/f8a_worker/ && \
     pip3 install -r requirements.txt
 
-RUN pip3 install git+https://github.com/fabric8-analytics/fabric8-analytics-utils.git@${F8A_UTILS_VERSION}
-RUN pip3 install git+https://git@github.com/fabric8-analytics/fabric8-analytics-version-comparator.git#egg=f8a_version_comparator
 COPY alembic.ini hack/run-db-migrations.sh ${ALEMBIC_DIR}/
 COPY alembic/ ${ALEMBIC_DIR}/alembic
-COPY selinon-2.0.0.tar.gz ${HOME}
-RUN pip3 install ${HOME}/selinon-2.0.0.tar.gz
```
Comment (arajkumar) [inline Dockerfile:29]: This has been already addressed by the other PR #911

File: requirements.in [hunk 2]
```diff
@@ -1,26 +1,65 @@
-# normally pulled in by kombu, but version 2.2.0 is broken
-amqp<=2.1.4
-# also handles Celery as a requirement
-selinon[celery]
-sqlalchemy
-psycopg2
-lxml
-beautifulsoup4
-# We install a patched version from dnf as we cannot use requests from PyPI - we need own certificates
-#requests
```
Comment (arajkumar) [inline requirements.in:10]: Can you please audit whether all of these direct dependencies are really used? I suspect no.

File: requirements.in [hunk 3]
```diff
@@ -1,26 +1,65 @@
...
+unidiff
+urllib3
+vine
 watchdog
-raven
+werkzeug
+xmltodict
+zipp
+git+https://github.com/fabric8-analytics/fabric8-analytics-utils.git
+git+https://github.com/fabric8-analytics/fabric8-analytics-version-comparator.git
```
Comment (arajkumar) [inline requirements.in:65]: pin version for these direct git based package installations.

File: tests/requirements.in [hunk 4]
```diff
@@ -1,10 +1,27 @@
+astroid
+atomicwrites
+attrs
+certifi
```
Comment (arajkumar) [inline tests/requirements.in:4]: Same here, are these many dependencies needed for unit test?

File: Dockerfile.rhel [hunk 5]
```diff
@@ -21,12 +20,8 @@ COPY requirements.txt /tmp/f8a_worker/
 RUN cd /tmp/f8a_worker/ && \
     pip3 install -r requirements.txt
 
-RUN pip3 install git+https://github.com/fabric8-analytics/fabric8-analytics-utils.git@${F8A_UTILS_VERSION}
-RUN pip3 install git+https://git@github.com/fabric8-analytics/fabric8-analytics-version-comparator.git#egg=f8a_version_comparator
 COPY alembic.ini hack/run-db-migrations.sh ${ALEMBIC_DIR}/
 COPY alembic/ ${ALEMBIC_DIR}/alembic
-COPY selinon-2.0.0.tar.gz ${HOME}
-RUN pip3 install ${HOME}/selinon-2.0.0.tar.gz
```
Comment (arajkumar) [inline Dockerfile.rhel:29]: This has been already addressed by the other PR #911

File: requirements.in (hunk 2 above)
Comment (yzainee) [reply requirements.in:10]: sqlalchemy, lxml are being used. not sure about amqp but its used for rabbit mq. Please verify this before removing them

Comment (jparsai) [reply requirements.in:10]: Yes, I have checked and removed unused packages.

Comment (arajkumar) [reply requirements.in:10]: I just tried with a tool called `pipreqs` which infers the dependency list by scanning the source code, it list only the following direct dependencies,

```
requests==2.24.0
pip==19.3.1
anymarkup==0.8.1
beautifulsoup4==4.9.1
boto3==1.14.34
botocore==1.17.34
celery==4.4.7
flexmock==0.10.4
git2json==0.2.3
GitPython==3.1.7
jsl==0.2.4
jsonschema==3.2.0
lxml==4.5.2
pytest==6.0.1
PyYAML==5.3.1
raven==6.10.0
requests_futures==1.0.0
selinon==1.1.0
semantic_version==2.8.5
SQLAlchemy==1.3.18
Werkzeug==1.0.1
```

Can you just try the above list without any version pinning(into requirements.in)? Of course you need to add https://github.com/fabric8-analytics/fabric8-analytics-utils and https://git@github.com/fabric8-analytics/fabric8-analytics-version-comparator.git.

Comment (arajkumar) [reply requirements.in:10]: There is no usage of https://git@github.com/fabric8-analytics/fabric8-analytics-version-comparator.git in the worker.

File: requirements.in [hunk 6]
```diff
@@ -1,26 +1,60 @@
...
+smmap
+soupsieve
+sqlalchemy
+toml<=0.9.4
+urllib3
+vine
+werkzeug
+xmltodict
+zipp
+git+https://github.com/fabric8-analytics/fabric8-analytics-utils.git@21ea911
```
Comment (arajkumar) [inline requirements.in:59]: Add egg=f8a_utils query param.

File: requirements.in [hunk 7]
```diff
@@ -1,26 +1,60 @@
...
+git+https://github.com/fabric8-analytics/fabric8-analytics-version-comparator.git@8a57ac7
```
Comment (arajkumar) [inline requirements.in:60]: This is not used in the worker.

Comment (yzainee) [reply requirements.in:60]: utils is used in worker.

Comment (arajkumar) [reply requirements.in:60]: I meant fabric8-analytics-version-comparator.git is not used.

File: requirements.in (hunk 2 above)
Comment (arajkumar) [reply requirements.in:10]: @yzainee if fabric8-analytics-version-comparator is a dependency of utils, then pip will automatically pull it right? Why should we add it here if it is not directly used by workers?

Comment (yzainee) [reply requirements.in:10]: Actually its not doing so. Never got a chance to investigate why. Probably a good time to check that out

Comment (arajkumar) [reply requirements.in:10]: @yzainee Fixed the transitive pull issue with the external packages. https://github.com/fabric8-analytics/fabric8-analytics-utils/pull/58

Comment (yzainee) [reply requirements.in:10]: +1

Comment (arajkumar) [reply requirements.in:10]: @jparsai Please make use of the latest utils. Also remove all unwanted dependencies using `pipreqs`.

File: requirements.in [hunk 8]
```diff
@@ -1,26 +1,22 @@
...
-boto3
-semantic_version
-radon==3.0.1
-watchdog
+jsl
+jsonschema
+lxml
+pyyaml
 raven
+radon<=3.0.1
```
Comment (arajkumar) [inline requirements.in:14]: Dev requirements should go to test/requiremnents.{in,txt}. Can you follow the check list created for this work? CC: @dgpatelgit

File: setup.py [hunk 9]
```diff
@@ -8,10 +8,19 @@
...
-        return fd.read().splitlines()
+    with open('requirements.in') as fd:
+        lines = fd.read().splitlines()
+        reqs, dep_links = [], []
+        for line in lines:
+            if line.startswith('git+'):
+                dep_links.append(line)
+            else:
+                reqs.append(line)
+        return reqs, dep_links
```
Comment (arajkumar) [inline setup.py:19]: This is incorrect, can you refer https://github.com/fabric8-analytics/fabric8-analytics-utils/pull/58? dependency_links is deprecated.

File: tests/requirements.in [hunk 10]
```diff
@@ -1,10 +1,6 @@
-datadiff
+pytest<=3.10.1
```
Comment (arajkumar) [inline tests/requirements.in:1]: Is there a reason for using pytest <=3.10.1?

Comment (jparsai) [reply tests/requirements.in:1]: Yes @arajkumar, This is the issue I was discussing with Dharmendra. For the PyTest version  >3.10.1, I am getting error **"Fixture "rdb" called directly. Fixtures are not meant to be called directly,"**, we looked into few online discussions about this, here is the link to one of them https://github.com/wolever/parameterized/issues/60 
To use grater version we will need to modify multiple test cases, so I have pointed it to 3.10.1 version.

Comment (arajkumar) [reply tests/requirements.in:1]: Ack!

**Review by arajkumar — APPROVED** [review]
lgtm

**Pull request merged**